KEYCLOAK-8349 Use TLS in the testsuite

slaskawi opened the pull request:
https://issues.jboss.org/browse/KEYCLOAK-8349

This PR changes the default behavior from non-TLS to TLS.

slaskawi committed:
KEYCLOAK-8349 KEYCLOAK-8659 Use TLS for all tests in the suite

stianst commented:
@keycloak-ci-bot test

hmlnarik commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/BasicSamlTest.java:
```diff
@@ -156,7 +156,7 @@ public void testNoPortInDestination() throws Exception {
     @Test
     public void testExplicitPortInDestination() throws Exception {
-        testWithOverriddenPort(Integer.valueOf(System.getProperty("auth.server.http.port")), Response.Status.OK, containsString("login"));
+        testWithOverriddenPort(Integer.valueOf(System.getProperty("auth.server.https.port")), Response.Status.OK, containsString("login"));
```
This needs to be conditioned on whether ssl tests are enabled or not.

slaskawi replied:
Done

slaskawi commented:
We also need to add a explicit test for HTTP

slaskawi commented:
`FixedHostnameTest#fixedHttpPort` could be a great example of a test, that explicitly disables TLS and hits HTTP endpoints. However at the moment, it's ignored (see [https://issues.jboss.org/browse/KEYCLOAK-8659](https://issues.jboss.org/browse/KEYCLOAK-8659)).

slaskawi commented:
@hmlnarik asked for a test run without TLS. It seems Travis is happy with this: https://travis-ci.org/slaskawi/keycloak/builds/447773742. Of course, once I finish all the bits and pieces, I'll rerun it.

slaskawi commented:
Rebasing... I will resolve this commend once done.

vmuzikar commented on testsuite/integration-arquillian/test-apps/servlets/src/main/java/org/keycloak/testsuite/adapter/servlet/SamlSPFacade.java:
```diff
@@ -90,7 +90,7 @@ private void handler(HttpServletRequest req, HttpServletResponse resp) throws Se
     private String getSamlRequest() {
         if (System.getProperty("auth.server.ssl.required", "false").equals("true")) {
-            return "jVLbTgIxEP2Vpu9Ly3JZbFgSlBhJUDeAPvhianeQJt127XS9%2FL1lwWiiokkfms6ZOZfpGGVlajFtwtYu4akBDOS1MhZFW8hp461wEjUKKytAEZRYTS8XIu1wUXsXnHKG7luOgyUi%2BKCdpWT6cT1zFpsK%2FAr8s1Zws1zkdBtCjYIx45Q0W4dBjIb9HoOqNu4NgFEyiyK1lbsBP8IHES6jIeZBmgpZCZVjH1rZzhcl584raF3ndCMNAiXzWU7ns%2Ft%2BOYKBgjTJUj5M%2Bln6kMjRSZak5aiX8ROuOB9EMBbRkH6Gz3bEBuYWg7QhpynvDhLeS%2FhwnaYinm7WicLuKCkOQk61LbV9PB7awx6E4mK9LpLie ...
+            return "jVJLbxshEL5Xyn9A3Ndg%2FNgN8lpyYkW1lDYr2%2B2hl4qw4xiJhQ3Dus2%2FD17HSqqoaQUHBN%2FM9xhmqBrbykUX924Njx1gJL8b61D2DyXtgpNeoUHpVAMoo5abxZdbKQZctsFHr72lp5KPwQoRQjTeUbI4H6%2B9w66BsIFwMBq%2BrW9Luo%2BxlYxZr5Xde4yyEAVn0LTWPwEwSpZJo3HqWH9C45%2FwyXjEVPLDAijbIKuh8ewslR1tUXLjg4bedEl3yiJQslqWdLX8Oa4LmGgQWS74NBvn4j5TxWWeiboY5fySa84nCYxV8mMO8FqO2MHKYVQullTw4STjo4xPt0LItIf5IAn7QUn1IuTKuNq4h48zuz% ...
```
During 7.3 testing I've discovered that this change broke `SAMLServletAdapterTest` when SSL is on for app server.
Usually, we test with SSL on for both app and auth server (or off for both). So imho as a temporary workaround we should hardcode this for this scenario. Currently, this change works only with SSL on for auth server and off for app server (or off for both).

slaskawi replied:
Yes, unfortunately, some of the tests are not prepared to run is TLS/non-TLS mode. For most of them I added `Assume.assumeTrue(AUTH_SERVER_SSL_REQUIRED)`, so perhaps I should do the same for `SAMLServletAdapterTest`?

hmlnarik replied:
Feel free to rebase on #5804 instead, it should help.

slaskawi replied:
Thanks for the heads up! I will once your PR gets into the master.

slaskawi commented:
Rebased

pedroigor commented:
@keycloak-ci-bot test

slaskawi commented:
Rebasing. Will let you guys know once I'm done.

slaskawi commented:
@keycloak-ci-bot test

slaskawi commented:
@pedroigor @hmlnarik @pdrozd @vramik Rebased. All builds are green, ready for review (and hopefully merge :smiley:)

pedroigor commented on integration/client-cli/client-registration-cli/src/main/java/org/keycloak/client/registration/cli/util/HttpUtil.java:
```diff
@@ -185,4 +189,22 @@ public static void setTruststore(File file, String password) throws CertificateE
         sslsf = new SSLConnectionSocketFactory(theContext);
     }
+
+    public static void setSkipCertificateValidation() {
```
If my comments to the other `HttpUtil` class are valid, they also apply here.

slaskawi replied:
I think that should be fine. Let's continue the discussion on the comment above.

pedroigor commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/AuthServerTestEnricher.java:
```diff
@@ -322,7 +334,50 @@ public void initializeTestContext(@Observes(precedence = 2) BeforeClass event) {
+                Administration administration = new Administration(client);
+                Operations operations = new Operations(client);
+
+                if(!operations.exists(Address.coreService("management").and("security-realm", "UndertowRealm"))) {
```
If we want to be future proof maybe we should consider Elytron configuration instead of using legacy configuration to enable TLS. Optional though, I'm OK too if you want to stick with legacy config.

slaskawi replied:
Since that piece is basically copy-pasted (and works fine on both EAP and WF), I would prefer to leave it as is. But thanks a lot for suggestion!

pedroigor commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/client/KeycloakTestingClient.java:
```diff
@@ -42,6 +45,11 @@
             resteasyClientBuilder.connectionPoolSize(10);
+            if (serverUrl.startsWith("https")) {
+                // Disable PKIX path validation errors when running tests using SSL
+                HostnameVerifier hostnameVerifier = (hostName, session) -> true;
```
Isn't this the same as hostnameVerification(ResteasyClientBuilder.HostnameVerificationPolicy.WILDCARD) ?

slaskawi replied:
Yes, fixed. Thanks!

pedroigor commented:
@slaskawi, glad to see that PR back to the discussion. Great addition to our test suite. Just a few comments that once solved/answered I'm OK to move forward and approve your changes.

hmlnarik reviewed (dismissed):
@slaskawi I appreciate the work done on this PR but SAML tests seem broken without TLS:

```
$ mvn -f testsuite/integration-arquillian/tests/base/pom.xml -Dtest=org.keycloak.testsuite.saml.** -nsu  test -Dauth.server.ssl.required=false
[...]
Tests run: 48, Failures: 34, Errors: 0, Skipped: 1
```
The same result is for Wildfly run:
```
mvn -f testsuite/integration-arquillian/tests/base/pom.xml -Dtest=org.keycloak.testsuite.saml.** -nsu clean test -Dauth.server.ssl.required=false -Pauth-server-wildfly
```

Without `-Dauth.server.ssl.required=false` the tests pass.

This needs to be fixed before merging.

OTOH I tried UserTest for OIDC and that works both with and without SSL and for both undertow and wildfly.

stianst commented on integration/client-cli/client-registration-cli/src/main/java/org/keycloak/client/registration/cli/commands/AbstractAuthOptionsCmd.java:
```diff
@@ -151,6 +150,8 @@ protected void setupTruststore(ConfigData configData, CommandInvocation invocati
                 throw new RuntimeException("Failed to load truststore: " + truststore, e);
             }
+        } else {
+            HttpUtil.setSkipCertificateValidation();
```
We shouldn't automatically skip certificate validation. Rather there should be an option to explicitly enable that if required.

slaskawi replied:
Granted and fixed.

I decided to completely change my strategy for the CLI tests. I reverted most of my changes and now use HTTP (and not HTTPS) port. Note, that the CLI tests explicitly test TLS/non-TLS scenario (for example in `KcAdmTruststoreTest#testTruststore`) so both, TLS and non-TLS scenarios are covered.

stianst commented on testsuite/integration-arquillian/test-apps/cors/angular-product/src/main/webapp/keycloak.json:
```diff
@@ -1,8 +1,9 @@
   "realm" : "cors",
   "realm-public-key" : "MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCrVrCuTtArbgaZzL1hvh0xtL5mc7o0NqPVnYXkLvgcwiC3BjLGw1tGEGoJaXDuSaRllobm53JBhjx33UNv+5z/UMG4kytBWxheNVKnL6GgqlNabMaFfPLPCF8kAgKnsi79NMo+n6KnSY8YeUmec/p2vjO2NjsSAVcWEQMVhJ31LwIDAQAB",
-  "auth-server-url" : "http://localhost-auth:8180/auth",
+  "auth-server-url" : "https://localhost-auth:8543/auth",
```
Doesn't this effectively force https for adapters?

slaskawi replied:
Good eye! I believe I should use Maven filtering there and set `http`/`https` and port depending on SSL setting. Let me fix that.

hmlnarik replied:
I believe this should be rather done by the deployment processor instead of Maven static build. See https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/DeploymentArchiveProcessor.java#L170-L183

slaskawi replied:
Unfortunately this is not an ordinary deployment, but rather a test app. It seems that our test apps are deployed as modules in WF and that means that we do not trigger `DeploymentArchiveProcessor`s on it. 

This one is heavily connected to @stianst's comment below. @stianst suggests that we should not disable trust manager and use proper trust store instead. So if we had a single point where we could inject a truststore into test applications, that would be lovely.

So I need an advice from you two (@hmlnarik and @stianst). Shall I try to refactor the existing applications to regular deployments? If so, it will take some time to do it (I bet the test suite will break up again). I see some benefits of doing this but having an opinion of someone more experienced might be helpful here.

hmlnarik replied:
This app is deployed in `org.keycloak.testsuite.adapter.example.cors.CorsExampleAdapterTest.angularCorsProductExample()` Is `DeploymentArchiveProcessor` somehow not followed here?

hmlnarik replied:
I think the issue is that the current `DeploymentArchiveProcessor` does not reflect angular path (i.e. no WEB-INF/ but webapp/ instead)

stianst replied:
I don't have much knowledge about the testing of adapters at all. This is probably a conversation for the team mailing list.

stianst commented on testsuite/integration-arquillian/test-apps/cors/angular-product/src/main/webapp/keycloak.json:
```diff
@@ -1,8 +1,9 @@
-  "auth-server-url" : "http://localhost-auth:8180/auth",
+  "auth-server-url" : "https://localhost-auth:8543/auth",
   "ssl-required" : "external",
+  "disable-trust-manager": true,
```
We shouldn't disable trust manager. One of the main points on SSL testing is that we make sure certs are verified properly and we don't end up allowing untrusted certificates automatically. Allowing untrusted certificates should only be possible if explicitly enabled.

stianst reviewed (commented):
I've only had time to have a very quick glance at this, but I have some concerns:

* I see several places where https is hard-coded. Especially in keycloak.json for adapters.
* I see several places where trust-manager/host-verification is disabled. One of the main points for SSL testing is to make sure certificates are always checked by default.

slaskawi commented on integration/client-cli/admin-cli/src/main/java/org/keycloak/client/admin/cli/util/HttpUtil.java:
```diff
@@ -262,6 +266,24 @@ public static void setTruststore(File file, String password) throws CertificateE
     }
 
+    public static void setSkipCertificateValidation() {
+        if (!tlsWarningEmitted.getAndSet(true)) {
```
I believe it should be implemented this way. The `HttpUtil` has a bunch of static methods and I can easily imagine a situation where you call its methods in many threads. In such a case, you would see this error emitted twice (or even more times). Of course, I'd like to avoid such situation, therefore, using an `AtomicBoolean` is the easiest way to avoid it.

vmuzikar reviewed (changes requested):
@slaskawi I ran the Admin Console UI tests with this branch.

IE is failing completely as it doesn't allow bypassing insecure (self-signed) SSL certificates. So it'd be best to disable SSL by default for this browser.

Chrome and FF works with one exception: `LoginSettingsTest.requireSSLAllRequests`. This should be fixed before merge. Maybe I could help with this.

vmuzikar commented:
@slaskawi Sure, I'm looking into the failing UI test.

Unfortunately, there are more failing tests.

**SocialLoginTest** This is failing for obvious reason. The fix seems to be just matter of changing the allowed redirect URI at the IdP side. IdP's OAuth protocols don't seem to mind the untrusted cert. Testing this solution now...

**Arquillian tests for Node.js adapter** This is a separate test module and is used temporarily anyway (until we have better coverage with the native Node.js tests). Since we don't yet test Node.js adapter with SSL ([KEYCLOAK-4909](https://issues.jboss.org/browse/KEYCLOAK-4909)), the fix should be fairly simple - just disabling SSL for the whole test module using POM.

I'll look into it on Monday.

I suspect there'll be more broken tests that are not executed in the test pipeline. However, currently I've focused only on tests I'm responsible for (i.e. I know how to run them :D).

One more thing I wanted to discuss. I suppose we don’t have the option to disable SSL on test class basis, correct? Something like “if you detect IE, don’t use SSL” or "for this test class, don't use SSL".
I know this can be done on demand on the CLI level while executing mvn test, but it’d be good to have the option to disable SSL by default in some cases.
The only solution I see is to do it in POM. But this would affect the whole test run, not just one test or so.

slaskawi commented:
@vmuzikar Thanks a lot for checking this PR against the UI tests!

Could you please tell me why those tests are not running in the pipeline by default? My suspicion is that those are the "old" testsuite and shall be migrated. If my understanding is correct, I think the best moment for adjusting those tests is the moment of migrating them to the "new" testsuite. What do you think?

> One more thing I wanted to discuss. I suppose we don’t have the option to disable SSL on test class basis, correct? Something like “if you detect IE, don’t use SSL” or "for this test class, don't use SSL".
I know this can be done on demand on the CLI level while executing mvn test, but it’d be good to have the option to disable SSL by default in some cases.
The only solution I see is to do it in POM. But this would affect the whole test run, not just one test or so.

The good news is that I haven't disabled the HTTP endpoint in the server. It's still in service. So all you need to do is to tell the OAuth Client to use it. You may look into `TLSTest` how to do it (this is a test where I'm explicitly checking if you can use HTTP even though HTTPS is enabled. Here's all you need to do:

```
public static final String AUTH_SERVER_ROOT_WITHOUT_TLS = "http://localhost:" + System.getProperty("auth.server.http.port", "8180") + "/auth";

// somewhere in your test
oauth.baseUrl(AUTH_SERVER_ROOT_WITHOUT_TLS);
// do your stuff
```

slaskawi commented:
@hmlnarik Answering your comment:

> @slaskawi I appreciate the work done on this PR but SAML tests seem broken without TLS:

I've just fixed them. It turned out that I used wrong wrong port in `SAML_ASSERTION_CONSUMER_URL_SALES_POST` (and friends). 

BTW (and it's out of the scope, I'm just curious), why we are using port `8080` there? The app server operates on `8180`. I see this port matches the SAML configuration but wouldn't that be easier if we used a real port there?

slaskawi commented:
@vmuzikar hmmm... At the moment, the situation is pretty hard to tackle. My suggestion would be to figure out the easiest way we could get those tests fixed and then get back to them once this PR is merged and all UI tests are running in the pipeline. Then, we could try to think how to do it better.

So if you agree, my suggestion would be to make the tests running by adding proper assumptions (`Assume.assumeFalse(AUTH_SERVER_SSL_REQUIRED)` for example) to the tests. Once everything is merged, we could get back to those parts and try to eliminate the assume statements. Such tests could be easily found by grepping the code and looking for "Assume False" statements.

slaskawi commented:
@stianst Thanks a lot for the review! Here are some answers to your comments:

> I see several places where https is hard-coded. Especially in keycloak.json for adapters.

Fixed. I switched to on-the-fly adapter configuration modifications in `DeploymentArchiveProcessorUtils`. Thanks @hmlnarik for the suggestion.

> I see several places where trust-manager/host-verification is disabled. One of the main points for SSL testing is to make sure certificates are always checked by default.

In most of the places, this has been fixed by the `DeploymentArchiveProcessorUtils` add-on. There are 2-3 places where I'm struggling with proper Trust Store implementation (e.g. a secured deployment invoked from CLI). I hope I will get them working in relatively short time and if not, I'll leave those 2-3 places until we find some proper solution.

vmuzikar committed:
UI and Node.js adapter tests fixes

vmuzikar commented:
@slaskawi I've created a PR https://github.com/slaskawi/keycloak/pull/2 (to your branch) with fixes for the UI tests and Node.js adapter tests.

IE is not fixed, though. For now, I've just added a note into HOW-TO-RUN to disable SSL while using IE. That was the simplest "solution". There maybe a better one (something like dismissing the warning dialog in IE) but this will most likely bring new stability problems so I don't like to drag this into this PR.

hmlnarik commented:
> BTW (and it's out of the scope, I'm just curious), why we are using port `8080` there? The app server operates on `8180`. I see this port matches the SAML configuration but wouldn't that be easier if we used a real port there?

That is because of 8080 being replaced here:
https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/DeploymentArchiveProcessor.java#L146-L161

slaskawi commented:
> @slaskawi I've created a PR slaskawi#2 (to your branch) with fixes for the UI tests and Node.js adapter tests.

Thank you @vmuzikar! I've integrated your changes into this PR.

mposolda commented:
@keycloak-ci-bot test

hmlnarik commented:
Much better this time yet few failures need investigation:
```bash
mvn clean install -Pauth-server-wildfly -fae -Dauth.server.ssl.required=false -f testsuite/integration-arquillian/pom.xml
[...]
Tests run: 2216, Failures: 7, Errors: 0, Skipped: 128
```
Failed tests are KcSamlBrokerTest, KcSamlSignedBrokerTest and KcSamlSignedDocumentOnlyBrokerTest: 
[org.keycloak.testsuite.broker.KcSamlBrokerTest.txt](https://github.com/keycloak/keycloak/files/2820675/org.keycloak.testsuite.broker.KcSamlBrokerTest.txt)
[org.keycloak.testsuite.broker.KcSamlSignedBrokerTest.txt](https://github.com/keycloak/keycloak/files/2820676/org.keycloak.testsuite.broker.KcSamlSignedBrokerTest.txt)
[org.keycloak.testsuite.broker.KcSamlSignedDocumentOnlyBrokerTest.txt](https://github.com/keycloak/keycloak/files/2820677/org.keycloak.testsuite.broker.KcSamlSignedDocumentOnlyBrokerTest.txt)

I'm going to run the tests with -Papp-server-wildfly now

hmlnarik commented:
With -Papp-server-wildfly, SecuredDeploymentsAdapterTest failed in addition to those above.

slaskawi commented:
> Much better this time yet few failures need investigation:

@hmlnarik Should be much better now.

hmlnarik commented:
4 methods of `KcSamlSignedBrokerTest` test are failing for 
`mvn clean install -Pauth-server-wildfly,app-server-wildfly -fae -Dapp.server.ssl.required=false -Dauth.server.ssl.required=false -f testsuite/integration-arquillian/pom.xml`
The rest has been fixed.

slaskawi commented:
> 4 methods of KcSamlSignedBrokerTest test are failing

Fixed @hmlnarik! Thanks for noticing. I forgot to push this part :man_facepalming:

vmuzikar commented on testsuite/integration-arquillian/HOW-TO-RUN.md:
```diff
@@ -475,38 +480,17 @@ To use a mobile browser you need to create a virtual device. The most convenient
 * **Supported mobile OS version:** iOS 11.x
 * **Run with:** `mvn clean test -Pios -Dappium.deviceName=device_name` where the device name is your device identification (e.g. `iPhone X`)
 
-## Run X.509 tests
```
@slaskawi One last thing. We shouldn't completely remove the instruction for X.509 tests. They are still excluded by default (they cannot run on Undertow).
The only change should be removing {{auth.server.ssl.required}} property from the instructions.

slaskawi replied:
Of course I can do this but to be honest, after enabling TLS by default, the procedure is quite standard (just like for any other test):

```
mvn -f ./testsuite/integration-arquillian/pom.xml clean install -Pauth-server-wildfly -Dtest="org.keycloak.testsuite.hok.HoKTest"
```

So I believe it doesn't bring any new information to the table. That's why I decided to erase it.

vmuzikar replied:
Ok, in this case I'm for removing:
https://github.com/keycloak/keycloak/blob/08a3f85e6da57de15945c0ae5aa7c9bf72d45206/testsuite/integration-arquillian/tests/base/pom.xml#L45
This is what makes the X.509 tests disabled no matter what. ;)

slaskawi replied:
It seems those tests are broken. Did we ever executed them (apart from manual execution)? If the answer is `no`, perhaps we could implement it in a separate JIRA?

vmuzikar replied:
I tested them yesterday with your branch - no failures. :)
You need to execute them with `-Pauth-server-wildfly`. Without it it fails.

slaskawi replied:
Ok, so we reached conclusion. It turned out that x509 tests require some small adjustments. They can only run properly with PhantomJS browser (it supports self-signed certificates out of the box). 

I reverted relevant to x509 tests paragraphs from `HOW-TO-RUN.md` and created [KEYCLOAK-9512](https://issues.jboss.org/browse/KEYCLOAK-9512) to implement all required changes.

vmuzikar reviewed (approved):
@slaskawi Thanks!

hmlnarik commented:
@keycloak-ci-bot test
